### DISCUSSION TITLE
Dynamic Resources

### DISCUSSION THREAD

**Discussion by multimeric** [ropensci/targets#931] [Help]
I have a set of files I want to analyse, which have been produced by `tar_files`. However, these files vary wildly in size, and therefore in memory required when I load these into memory:
```R
list(
  tar_files(
    name = sample_dirs,
    command = ...
  ),
  tar_target(
    name = loaded_count_db,
    command = ...,
    resources = tar_resources(
      future = tar_resources_future(
        plan = tweak(
          batchtools_slurm,
          resources = list(memory = 1024 * 10),
          template="slurm"
        )
      )
    )
  )
)
```
If targets supported a dynamic resource selection, I could define an upstream task that queried the size of each file and use that to estimate the memory usage, but this doesn't seem possible. The only solution I can think of at the moment is to duplicate the processing task several times, one for each group of files, and then define an upstream task that segregates the data based on its size, e.g.
```R
list(
  tar_files(
    name = get_big_samples,
    command = ...
  ),
  tar_files(
    name = get_medium_samples,
    command = ...
  ),
  tar_files(
    name = get_small_samples,
    command = ...
  ),
  tar_target(
    name = load_big_samples,
    command = ...,
    resources = tar_resources(
      future = tar_resources_future(
        plan = tweak(
          batchtools_slurm,
          resources = list(memory = 1024 * 10),
          template="slurm"
        )
      )
    )
  ),
  tar_target(
    name = load_medium_samples,
    command = ...,
    resources = tar_resources(
      future = tar_resources_future(
        plan = tweak(
          batchtools_slurm,
          resources = list(memory = 1024 * 5),
          template="slurm"
        )
      )
    )
  ),
  tar_target(
    name = load_small_samples,
    command = ...,
    resources = tar_resources(
      future = tar_resources_future(
        plan = tweak(
          batchtools_slurm,
          resources = list(memory = 1024),
          template="slurm"
        )
      )
    )
  )
)
```

This seems a bit inelegant however. Is there a better way to do this?

**Comment by wlandau** [comment]
Yeah, resources are completely static. You may already be doing this, but one of the easiest ways to reduce the burden of in-session memory is to opt into transient memory and garbage collection as described at https://books.ropensci.org/targets/performance.html. Otherwise, `tarchetypes::tar_files_read()` static branching with `tarchetypes::tar_eval()` and/or a custom [target factory](https://wlandau.github.io/targetopia/contributing.html#target-factories) could condense that second target list down to fewer lines of code if you are willing to do a bit of metaprogramming.
